webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,151 @@
+            if (msg['sync_id']) {
+                if (!msg['complete']) {
+                    this.#syncStarted = msg['sync_id'];
+                } else {
+                    this.#syncComplete = msg['sync_id'];
+                    this.advanceSubscription();
+                }
+            } else {
+                // Ignore all messages between when we send a new subscription and when the
+                // sync-start message for that subscription arrives.  These are the online
+                // updates for a subscription we no longer care about.
+                if (this.#syncSent !== this.#syncStarted) return;
+
+                forEach(msg, (val, k) => {
+                    if (k.includes('_deleted')) {
+                        const stream_key = trimEnd(k, '_deleted') as Streamable;
+                        const deleted_keys = decode_keys(val as string);
+                        this.#curEntity?.[stream_key].keyCache.delete_msg(deleted_keys);
+                        onDelete(stream_key, deleted_keys);
+                    } else {
+                        this.#curEntity?.[StreamEntityMap[k]].keyCache.upsert([val.id], val.seq);
+                        onUpsert(msg);
+                    }
+                });
+
+            }
+
+        };
+    }
+
+    async retry(): Promise<boolean> {
+        try {
+            const backoff = backoffs[this.#retries];
+            this.#retries += 1;
+            await sleep(backoff);
+            return true;
+        } catch {
+            return false;
+        }
+    }
+
+    #sameAsCur(entity: Record<Streamable, Entity>): boolean {
+        if (!this.#curEntity) return false;
+        let same = true;
+        forEach(entity, (val, k) => {
+            if (!this.#curEntity?.[k as Streamable].spec.equals(val.spec)) same = false;
+        });
+        return same;
+    }
+
+    #sendSpec(entity?: Record<Streamable, Entity>): void {
+        if (!entity || this.#sameAsCur(entity)) return;
+
+        this.#numSyncs += 1;
+        const sync_id = this.#numSyncs.toString();
+        const spec = reduce(entity, (spec, ent, k) => {
+            spec['known'][k] = ent.keyCache.known();
+            spec['subscribe'][k] = {
+                ...ent.spec.toWire(),
+                since: ent.keyCache.maxSeq(),
+            };
+            return spec;
+        }, { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, any>);
+
+        this.#ws.send(JSON.stringify(spec));
+        this.#curEntity = entity;
+        this.#syncSent = sync_id;
+    }
+
+    advanceSubscription(): void {
+        if (this.#ws.readyState !== this.#ws.OPEN) return;
+        if (!this.#curEntity && !this.#entities) return;
+
+        let spec: Record<Streamable, Entity> | undefined;
+        if (!this.#syncSent) {
+            if (this.#curEntity) {
+                spec = this.#curEntity;
+            } else {
+                spec = this.#entities.shift();
```

Comment:
can this line of code fire more more than once?

Like I'm sure the caller can request multiple subscriptions, but I'm not sure I see how the second one would ever become active